session/session.go

```diff
@@ -90,6 +108,152 @@ func Open(config sfdc.Configuration) (*Session, error) {
 	return session, nil
 }
 
+func openDeviceSession(config sfdc.Configuration) (*Session, error) {
+	request, err := buildDeviceAuthenticationFlowInitiationRequest(config.Credentials)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build ")
+	}
+
+	initResp, err := makeDeviceAuthenticationFlowInitiationRequest(request, config.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	var tokenResp *accessTokenResponse
+	var done = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Duration(initResp.Interval+1) * time.Second):
+				accessTokenReq, err := buildDeviceAuthenticationFlowAccessRequest(config.Credentials, initResp)
+				if err != nil {
+					log.Fatalf("failed to build device authentication flow access request: %s", err)
+				}
+
+				// Make authorization request
+				var tokenErrResp *deviceAccessTokenErrorResponse
+				if tokenResp, tokenErrResp, err = makeDeviceAuthenticationFlowAccessTokenRequest(accessTokenReq, config.Client); err != nil {
+					log.Fatalf("error polling for device access token: %s", err)
+				}
+
+				// Success!
+				if tokenResp != nil {
+					close(done)
+					return
+				}
+
+				if tokenErrResp != nil && tokenErrResp.Error == AuthorizationPendingErrorCode {
+					log.Printf("Authorization pending. Please enter code '%s' at %s to authorize application.", initResp.UserCode, initResp.VerificationURI)
+					log.Printf("will attempt to authorize again in %d seconds", initResp.Interval)
+				} else {
+					log.Printf("unhandled authorization response error: %s: %s", tokenErrResp.Error, tokenErrResp.ErrorDescription)
+				}
+			}
+		}
+	}()
+
+	// Wait for successful authorization response.
+	<-done
+
+	session := &Session{
+		response: tokenResp,
+		config:   config,
+	}
+
+	return session, nil
+}
+
+func buildDeviceAuthenticationFlowInitiationRequest(creds *credentials.Credentials) (*http.Request, error) {
+	oauthURL := creds.URL() + oauthEndpoint
+
+	body, err := creds.Retrieve()
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := http.NewRequest(http.MethodPost, oauthURL, body)
+
+	if err != nil {
+		return nil, err
+	}
+
+	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	request.Header.Add("Accept", "application/json")
+	return request, nil
+}
+
+func makeDeviceAuthenticationFlowInitiationRequest(request *http.Request, client *http.Client) (*deviceAuthenticationFlowInitiationResponse, error) {
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("session response error: %d %s", response.StatusCode, response.Status)
+	}
+	decoder := json.NewDecoder(response.Body)
+	defer response.Body.Close()
+
+	var sessionResponse deviceAuthenticationFlowInitiationResponse
+	err = decoder.Decode(&sessionResponse)
```

Comment:
since `decoder` variable is used only once we can reduce 195 line and this one to:
```go
err = json.NewDecoder(response.Body).Decode(&sessionResponse)
```